### PR TITLE
[c10d] Make reduce as a custom op

### PR DESCRIPTION
**Pull request by alanwaketan**
Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* #79779
* #79777
* #79691
* #79688
* #79687
* __->__ #79686
* #79683
* #79669
* #79582

Summary:
This patch makes reduce as a custom op such that it's dispatcher
passable. It's one part of the effort to route comm ops to the dispatcher
such that tracing mechanisms that relies on the dispatcher can trace them,
e.g., LazyTensor and AOTAutograd.

Test Plan:
python test/distributed/test_c10d_nccl.py -k test_reduce_ops
python test/distributed/test_c10d_gloo.py -k test_reduce_basics
...and other existing distributed tests.

### CODE REVIEW

File: torch/csrc/distributed/c10d/init.cpp [hunk 1]
```diff
@@ -1060,22 +1060,26 @@ that adds a prefix to each key inserted to the store.
...
+                return ::c10d::ops::reduce(self, tensors, opts);
+              },
               py::arg("tensors"),
               py::arg("opts") = ::c10d::ReduceOptions(),
               py::call_guard<py::gil_scoped_release>())
 
           .def(
               "reduce",
-              [](::c10d::ProcessGroup& pg,
+              [](const c10::intrusive_ptr<::c10d::ProcessGroup>& pg,
```
Comment (mrshenli) [inline torch/csrc/distributed/c10d/init.cpp:1074]: ditto on naming

**Review by mrshenli — APPROVED** [review]
LGTM

File: torch/csrc/distributed/c10d/init.cpp (hunk 1 above)
Comment (alanwaketan) [reply torch/csrc/distributed/c10d/init.cpp:1074]: Let me follow up on #80246. Also, you forgot to approve this pull request. : )

**Pull request closed without merging**